Netable/Netable/GraphQLRequest.swift

```diff
@@ -8,21 +8,65 @@
 
 import Foundation
 
+/// Specify how to retrieve the GraphQL query for a request
+public enum GraphQLQuerySource {
+    /// Autogenerate resource file from dynamic class name (risky)
+    /// This functionality checks the name of the request and uses `type(of: self)` to find a .graphql file with the same name in the bundle.
+    case autoResource
+
+    /// An arbitrary file URL
+    case file(URL)
+
+    /// The full text string of the query
+    case literal(String)
+
+    /// The name of a .graphql file in the app bundle
+    /// You don't need to include the file extension when specifying a resource.
+    case resource(String)
+}
+
+/*
+*   This helper protocol provides a set of shared methods used by `GraphQLQuery` and `GraphQLMutation`.
+*   You'll likely want to sub-class those protocols instead of this one.
+*/
 public protocol GraphQLRequest: Request {
     associatedtype Parameters = [String: String]
 
-    func getGraphQLQueryFileContents() -> String
+    var source: GraphQLQuerySource { get }
 }
 
 public extension GraphQLRequest {
     var method: HTTPMethod { HTTPMethod.post }
 
     var path: String { "" }
 
-    func getGraphQLQueryFileContents() -> String {
-        guard let resource = "\(type(of: self))".split(separator: ".").last,
-            let resourcePath = Bundle.main.path(forResource: String(resource), ofType: "graphql"),
-            let params = try? String(contentsOfFile: resourcePath) else {
+    func getGraphQLQueryContents() -> String {
+        switch source {
+        case .autoResource: return getAutoResourcedContents()
+        case .file(let url): return getFileContents(url)
+        case .literal(let literal): return literal
+        case .resource(let bundleResource): return getResourcedContents(bundleResource)
+        }
+    }
+
+    private func getFileContents(_ url: URL) -> String {
+        guard let params = try? String(contentsOf: url) else {
```

Comment:
Not something we need to fix right now, but ideally we shouldn't be doing a synchronous file read here (or any of the other cases where were are opening a file to get this), since we don't really want to ever block the main thread on I/O, even hopefully fast I/O.

I don't think we should bother doing anything about it until we've added properly concurrency support to the library (#77), it's be too much work for minimal benefit, but when we DO add concurrency support we should flag this as something that should specifically be converted as well rather than leaving it synchronous.

Reply:
Added a note in the async/await issue to address this